No names for circuit library gate definitions

Cryoris opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

In https://github.com/Qiskit/qiskit/pull/15260 we realized that we're currently making it very easy for users to violate one of Qiskit's assumptions: standard gate names should be unique. Currently, most gates in the circuit library use the name of the gate as name of the definition circuit. Meaning workflows like the following make it easy to cut yourself:
```python
gate = RXGate(0.1)  # or some other gate from the library
sharp_bit = gate.definition.to_gate() # a gate called "rx", but not an RXGate instance
```

This commit sets the optional name of the gate definitions to `None`, alleviating this issue a bit. This also fixes the problems in https://github.com/Qiskit/qiskit/pull/15260.

Cryoris committed:
Make Qiskit less sharp

Make it less likely to accidentally end up with an opaque instruction that has the name of a standard gate, not naming the definition of a gate with the gate name. E.g. RXGate(..).definition.name shouldn't be "rx".

Cryoris committed:
Merge branch 'main' into noname

jakelishman reviewed (commented):
Imo we should put an "upgrade" release note mentioning this, even though it's not _expected_ to be user facing, because there is a reason we're doing it.

Cryoris committed:
Fix single controlled test

This test just happened to pass since the opaque gate we controlled was called "rz", triggering the "rz" path in add_control -- but in fact the gate represented a "p" gate. This is exactly the sort of facilitated foot-shooting this PR wants to avoid.

Cryoris committed:
add reno

alexanderivrii commented on qiskit/circuit/library/generalized_gates/gms.py:
```diff
@@ -158,7 +158,7 @@ def __init__(
         thetas = self.params[0]
         q = QuantumRegister(self.num_qubits, name="q")
-        qc = QuantumCircuit(q, name=self.name)
+        qc = QuantumCircuit(q)
```
I am not so worried about names for _non-standard gates_ (such as "gms" here), so I am fine either way. But if you do want to do this change for all gates, what about also `LinearPauliRotationGate`, `PiecewiseLinearPauliRotationGate`, `PiecewisePolynomialPauliRotationGate`, and `UnitaryGate`?

Cryoris replied:
Yeah I think it would be better to do this for _all_ the gates, there's no reason that the definition should have the same name 👍🏻

alexanderivrii commented on test/python/circuit/test_controlled_gate.py:
```diff
@@ -1540,21 +1540,13 @@ class TestSingleControlledRotationGates(QiskitTestCase):
-    @data((gu1, cgu1), (grx, cgrx), (gry, cgry), (grz, cgrz))
-    @unpack
-    def test_single_controlled_rotation_gates(self, gate, cgate):
+    @data(gu1, grx, gry, grz)
+    def test_single_controlled_rotation_gates(self, gate):
         """Test the controlled rotation gates controlled on one qubit."""
+        as_circuit = QuantumCircuit(1)
+        as_circuit.append(gate, [0])
+        cgate = as_circuit.control(num_ctrl_qubits=2)
+
```
Technically this changes the test a bit, but the new way is the correct way to construct controlled quantum circuits, so I fully support this change.

alexanderivrii commented on releasenotes/notes/clib-definition-noname-75d0daf1355f4038.yaml:
```diff
@@ -0,0 +1,7 @@
+---
+upgrade_circuits:
+  - |
+    The name of the circuit produced as definitions of circuit library gates were set to ``None``.
```
I am not sure which words in this sentence should be singular-vs-plural ("names"? "circuits"? "definitions"? "gates"?) but I do believe that it should be either "The name ... [was or is] set" or "The names ... [were or are] set".

```suggestion
    The names of the circuits produced as definitions of standard circuit library gates are now set to ``None``.
```

alexanderivrii reviewed (commented):
Thanks @Cryoris, this looks good! I have double-checked that we no longer set circuit names for the definitions of _standard_ gates in the circuit library. A minor question is whether we should also do this for generalized gates (where this PR does this for some but not for all generalized gates). And a minor suggestion for the release note.

Cryoris committed:
Merge branch 'main' into noname

Cryoris committed:
remove some additional name collisions

Cryoris committed:
Update releasenotes/notes/clib-definition-noname-75d0daf1355f4038.yaml

Co-authored-by: Alexander Ivrii <alexi@il.ibm.com>

alexanderivrii reviewed (approved):
Thanks for the update, LGTM!
